machineconfiguration/v1/types.go

```diff
@@ -0,0 +1,389 @@
+	// Total number of ready machines targeted by the pool.
+	ReadyMachineCount int32 `json:"readyMachineCount"`
+
+	// Total number of unavailable (non-ready) machines targeted by the pool.
+	// A node is marked unavailable if it is in updating state or NodeReady condition is false.
+	UnavailableMachineCount int32 `json:"unavailableMachineCount"`
+
+	// Total number of machines marked degraded (or unreconcilable).
+	// A node is marked degraded if applying a configuration failed..
+	DegradedMachineCount int32 `json:"degradedMachineCount"`
+
+	// Represents the latest available observations of current state.
+	Conditions []MachineConfigPoolCondition `json:"conditions"`
+}
+
+// MachineConfigPoolStatusConfiguration stores the current configuration for the pool, and
+// optionally also stores the list of machineconfig objects used to generate the configuration.
+type MachineConfigPoolStatusConfiguration struct {
+	corev1.ObjectReference `json:",inline"`
+
+	// source is the list of machineconfigs that were used to generate the single machineconfig object specified in `content`.
+	// +optional
+	Source []corev1.ObjectReference `json:"source,omitempty"`
+}
+
+// MachineConfigPoolCondition contains condition information for an MachineConfigPool.
+type MachineConfigPoolCondition struct {
+	// Type of the condition, currently ('Done', 'Updating', 'Failed').
+	Type MachineConfigPoolConditionType `json:"type"`
+
+	// Status of the condition, one of ('True', 'False', 'Unknown').
+	Status corev1.ConditionStatus `json:"status"`
+
+	// LastTransitionTime is the timestamp corresponding to the last status
+	// change of this condition.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+
+	// Reason is a brief machine readable explanation for the condition's last
+	// transition.
+	Reason string `json:"reason"`
+
+	// Message is a human readable description of the details of the last
+	// transition, complementing reason.
+	Message string `json:"message"`
+}
+
+// MachineConfigPoolConditionType valid conditions of a machineconfigpool
+type MachineConfigPoolConditionType string
+
+const (
+	// MachineConfigPoolUpdated means machineconfigpool is updated completely.
+	// When the all the machines in the pool are updated to the correct machine config.
+	MachineConfigPoolUpdated MachineConfigPoolConditionType = "Updated"
+	// MachineConfigPoolUpdating means machineconfigpool is updating.
+	// When at least one of machine is not either not updated or is in the process of updating
+	// to the desired machine config.
+	MachineConfigPoolUpdating MachineConfigPoolConditionType = "Updating"
+	// MachineConfigPoolNodeDegraded means the update for one of the machine is not progressing
+	MachineConfigPoolNodeDegraded MachineConfigPoolConditionType = "NodeDegraded"
+	// MachineConfigPoolRenderDegraded means the rendered configuration for the pool cannot be generated because of an error
+	MachineConfigPoolRenderDegraded MachineConfigPoolConditionType = "RenderDegraded"
```

Comment:
Do folks need to distingish between render and node errors by type?  Or can we just have `Degraded` and set different messages?

Reply:
there's a _global_ Degraded status but there are 2 controllers playing here and we need to different fields to create a meaningful logic to report Degraded (as discussed with Abhinav long time ago).